### DISCUSSION TITLE
Using the Next 13 `app` directory on Netlify

### DISCUSSION THREAD

**Discussion by ascorbic** [opennextjs/opennextjs-netlify#1724] [Q&A]
If you update your site to Next 13, it will work on Netlify right away. All the stable features work out of the box with no changes required on your part when using the latest Netlify Next.js runtime. If you have manually installed the Netlify Next.js runtime in your package.json you will need to upgrade it, but for everyone else it will upgrade you automatically on your next deploy. However this does not include the new experimental `app` directory features. This is currently pre-release and should not be used for production.

While deploying `app` directory sites to Netlify is still experimental, it already supports many of the most interesting new features, including React Server Components and streaming HTML using the edge runtime. Pages that use the edge runtime are automatically deployed to Netlify Edge Functions.

This topic is for discussion of the use of `app` directory features, not general Next 13 topics. If you encounter a problem with an existing feature such as the `pages` directory or middleware, then open an issue rather than commenting here.

Update: you no longer need to install a special version of the runtime - the auto-installed version has appDir support enabled, though it is still experimental

## Known issues
- You cannot currently use `next.config.js` to rewrite to a page that uses the edge runtime. The workaround is to use middleware to rewrite, which works correctly.
- dynamicParams: false is not supported
- https://github.com/netlify/next-runtime/issues/1800
- https://github.com/netlify/next-runtime/issues/1798
- https://github.com/netlify/next-runtime/issues/1797
- https://github.com/netlify/next-runtime/issues/1796
- https://github.com/netlify/next-runtime/issues/1747
-

**Comment by rohanbhangui** [comment]
Seems to break with the latest version of nextjs 13.0.3. Getting an internal server error 500.

@ascorbic  should we continue to use the next js runtime at  4.29.1 or should we lock to  4.28.4-appdir.0 for now?

**Reply by ascorbic** [reply]
I will be putting a new version of the appdir branch out tomorrow

**Reply by rohanbhangui** [reply]
@ascorbic thanks! Appreciate it :)

**Reply by rohanbhangui** [reply]
@ascorbic QQ is there anyway to have the runtime auto upgrade only to the appDir versions (I know one usually uses `^` to indicate versions higher than this)

**Reply by ascorbic** [reply]
If you explicitly set the dependency to `"@netlify/plugin-nextjs": "appdir"` then that should do it

**Comment by kamalbennani** [comment]
Hello guys 👋 
So, I'm trying to deploy my NextJS 13 test app to Netlify (of course using appdir experimental feature) and weirdly enough all my routes navigation is turned into a hard page refresh 🤔 Is that expected? 
Here is the [preview](https://63b74bb355b7ef27f6491c29--aircall-hybrid.netlify.app/) of the app.
If you click on "Hello" text it should redirect you to the hello page using a hard page refresh.

**Reply by kamalbennani** [reply]
I'm assuming it's linked to this [PR](https://github.com/netlify/next-runtime/pull/1855)

**Reply by orinokai** [reply]
Yes, that's right @kamalbennani 
We are currently implementing a new method for handling RSC data, which will allow Next Link to get the data it requires for client-side navigation. Hold tight, it's en route!

**Reply by adshodgson** [reply]
Thanks, I am also having the same issue occur, excited for this patch!

**Reply by richardfwashington** [reply]
I too seem to be having the same issue

**Reply by sebastianhaiss** [reply]
I’m also seeing the same issue of complete page re-renders when navigating with the <Link> component. Looking forward for a fix as well! 🤓

**Reply by richardfwashington** [reply]
Just tried the app on Vercel, and seems to be fine.  So it does at least appear to be related to Netlify in particular.

**Reply by bentrynning** [reply]
Any update to fix this behaviour? 😊

**Comment by ben519** [comment]
Not sure if this is a Netlify issue or a problem with my own code, but I keep getting a 500 Internal Server error when trying to `fetch()` data. I've described my issue in detail on [StackOverflow](https://stackoverflow.com/questions/75222855/500-internal-server-error-fetching-data-with-next-js-app-deployed-on-netlify).

**Reply by ascorbic** [reply]
Hi. I've replied on SO, but looking at your logs you're getting the error "Error: Page changed from static to dynamic at runtime /team/1, see more here https://nextjs.org/docs/messages/app-static-to-dynamic-error"

**Reply by ben519** [reply]
This solved my issue. Thank you so much for the tips!

**Comment by ben519** [comment]
It seems that Next.js 13's [loading UI](https://beta.nextjs.org/docs/routing/loading-ui) doesn't work with Netlify (i.e. [React Suspense component](https://beta.reactjs.org/reference/react/Suspense).)

My suspense component works fine locally with `npm run dev` and `ntl dev`, but it doesn't work when deployed to Netlify. One clear way to see this is by using the [React developer tools Chrome extension](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en).

**`ntl dev`**
<img width="721" alt="Screenshot 2023-01-26 at 3 50 05 PM" src="https://user-images.githubusercontent.com/723099/214958666-48d4c399-acc8-45a7-8da2-1063f52a943f.png">

**[https://loading--dundermifflininc.netlify.app/team](https://loading--dundermifflininc.netlify.app/team)**
<img width="594" alt="Screenshot 2023-01-26 at 3 50 39 PM" src="https://user-images.githubusercontent.com/723099/214958687-b3459a61-a40a-40c9-a2d6-47c8c99411d9.png">

**Reply by ascorbic** [reply]
I'll take a look, thanks. Could be because we don't support streaming from the Node runtime?

**Reply by ben519** [reply]
Thanks! Sounds like a reasonable hypothesis, but a bit over my pay grade at this time.

**Reply by ClaytonTreder** [reply]
I have noticed this issue as well. Commenting to hopefully stay in the loop. Cheers!

**Comment by hk86** [comment]
Someone using appDir and trailingSlash with nextjs@13.1.6? Opening the url www.example.com/blog/article-a/ results in a redirect to www.example.com/blog/article-a

I've also deployed the same site to vercel and there trailingSlash is respected. I get no redirect.

Is this a known issue?

**Comment by sadhikari1r** [comment]
Issue: 500 Internal Server Error when deploying Next 13.4 with embed sanity studio on Netlify.

Adding an embed Sanity studio on Next 13.4. causes "500: Internal Server Error". The problem seems to be specifically related to the serverActions in Next.js 13.4, as the application works fine locally and on previous versions (e.g., 13.3) on the production.

Here is the log on Netlify Functions:

`
May 9, 12:02:35 PM: c0984344 ERROR  Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './server.edge' is not defined by "exports" in /var/task/node_modules/react-dom/package.jsonMay 9, 12:02:35 PM: c0984344 ERROR      at new NodeError (node:internal/errors:387:5)May 9, 12:02:35 PM: c0984344 ERROR      at throwExportsNotFound (node:internal/modules/esm/resolve:365:9)May 9, 12:02:35 PM: c0984344 ERROR      at packageExportsResolve (node:internal/modules/esm/resolve:649:3)May 9, 12:02:35 PM: c0984344 ERROR      at resolveExports (node:internal/modules/cjs/loader:554:36)May 9, 12:02:35 PM: c0984344 ERROR      at Function.Module._findPath (node:internal/modules/cjs/loader:594:31)May 9, 12:02:35 PM: c0984344 ERROR      at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1012:27)May 9, 12:02:35 PM: c0984344 ERROR      at /var/task/node_modules/next/dist/server/require-hook.js:180:36May 9, 12:02:35 PM: c0984344 ERROR      at Function.Module._load (node:internal/modules/cjs/loader:871:27)May 9, 12:02:35 PM: c0984344 ERROR      at Module.require (node:internal/modules/cjs/loader:1098:19)May 9, 12:02:35 PM: c0984344 ERROR      at require (node:internal/modules/cjs/helpers:108:18) {May 9, 12:02:35 PM: c0984344 ERROR    code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'May 9, 12:02:35 PM: c0984344 ERROR  }
`

As a temporary workaround, setting the environment variable "__NEXT_PRIVATE_PREBUNDLED_REACT=next" resolved the issue for others. However, Netlify does not allow this particular environment variable to be set due to the prefix of (_).

During investigation, I came across a discussion thread in the Netlify Support Forums that addresses a similar error: [Netlify Support Forums Thread](https://answers.netlify.com/t/server-edge-not-defined-error-on-nextjs-ssr-functions-cause-site-to-return-500-errors/91793).

Additionally, I noticed that there is a pull request in the Next.js GitHub repository that appears to address similar issue specifically for custom servers. It seems to have been fixed and merged on v13.4.0 but many others as well as I'm still getting the error. This might just be an issue with Next itself. Just wanted to point it for someone from Netlify to look into. [Next.js PR](https://github.com/vercel/next.js/issues/49169).

**Reply by nickytonline** [reply]
@sadhikari1r, we're currently upgrading to 13.4 in the Next.js runtime. You can follow along here, https://github.com/netlify/next-runtime/pull/2080

**Reply by nickytonline** [reply]
Everything was merged and is currently rolling out. It'll be available to everyone by the end of the week. See https://github.com/netlify/next-runtime/releases/tag/plugin-nextjs-v4.38.0

**Reply by j-hit** [reply]
How can I explicitly define that Netlify should use v4.38.0 next.js runtime?

**Reply by morgnism** [reply]
The latest runtime still doesn't resolve this issue. I have an app running on 13.4 and still get this error with group routes `i.e. app/(studio)/admin/[[...index]]/page.tsx` (current sanity studio path with catchall route). I assumed that there's no way to manually change Nextjs runtime in Netlify and that it's updated automatically to the latest version.

Is there something we're missing?

**Reply by morgnism** [reply]
@j-hit you should be able to via toml and nextjs plugin ([ref the plugin readme](https://github.com/netlify/next-runtime#from-npm)). However, this should be automatically updated via Netlify UI.

---
Just found [this related support article](https://answers.netlify.com/t/server-edge-not-defined-error-on-nextjs-ssr-functions-cause-site-to-return-500-errors/91793/70) where this problem is still present and being worked on.

**Comment by yukilun** [comment]
Hi! I am trying to deploy a next 13 app using 'app' router to Netlify. I added a[ 'not-found.js'](https://nextjs.org/docs/app/api-reference/file-conventions/not-found) to the 'app' directory to create my custom 404 page so that it will be shown when user visit any invalid URL .

I tried to create a repo for testing with just the starter file (npx create-next-app@latest). Without 'not-found.js', the Next default 404 page would be shown when I went to an invalid URL (https://647825f32dfd7e319f25f51d--regal-pika-910873.netlify.app/unkown):
![image](https://github.com/netlify/next-runtime/assets/102889935/7d12cb3a-ce24-4391-a872-e029f62a5b49)

However, when I added the 'not-found.js' to 'app' directory and visited an invalid URL, the custom 404 page could not be shown. Instead, "Internal Server Error" was shown (https://647826e083bb9900089436a2--regal-pika-910873.netlify.app/unkown):
![image](https://github.com/netlify/next-runtime/assets/102889935/12353d2d-1cb0-43f1-ade7-a93dbc0cc475)

I have also tried to deploy a next app using 'page' router and added '404.js' to 'page' directory and it works well(https://amazing-quokka-cb8229.netlify.app/unknown):
![image](https://github.com/netlify/next-runtime/assets/102889935/10d8a626-6bea-4cdf-b199-f2e910130726)

Is there any solution if I would like creating my custom page with 'app' directory? Thank you.

**Reply by MichaelTae** [reply]
Is there any word on this? I'm having the same issue and I can't seem to figure it out.

**Reply by SupratikRulz** [reply]
same here, I am looking out for a solution to this issue

**Comment by j-hit** [comment]
When deploying a Next.js 13.4.4 app to Netlify I had some issues.

At first I thought the deployment worked but I got the Netlify 404 page when I went to the site's URL.

In Netlify the build was not executed, only the post-processing stage was executed.
<img width="609" alt="image" src="https://github.com/netlify/next-runtime/assets/8820342/e7007a95-3d5c-481c-bf16-1249ed0257a2">

For it to work I just had to add the following to a `netlify.toml` file in the root of the repo.
```
[build]
  publish = ".next"
```

For full context my `.gitlab-ci.yml` file looks like this.
```yaml
stages:
  - build
  - deploy
build-app:
  stage: build
  image: node:16
  script:
    - npm install
    - npm run build
  artifacts:
    paths:
      - .next/
netlify:
  stage: deploy
  image: node:16
  script:
    - npm install -g netlify-cli
    - netlify deploy --site $NETLIFY_SITE_ID --auth $NETLIFY_AUTH_TOKEN --build --prod --dir=.next
  only:
    - main
```

**Comment by norberthnd** [comment]
Hi! Dynamic routes (generated from mdx files) do not work with netlify form actions.
If such routes are used in the form action field, it results in a timeout:
<img width="878" alt="CleanShot 2023-06-16 at 22 34 07@2x" src="https://github.com/netlify/next-runtime/assets/13472960/c96010f5-d1ce-4f2c-82d3-0c30e1242d42">

The form submit works if the same page is created (as a directory under the app).

**Reply by orinokai** [reply]
hi @norberthnd, we'd need to discuss a few more details to diagnose the problem here and it would be great if you could open a GitHub issue or get in touch at https://www.netlify.com/support/

**Comment by j-hit** [comment]
Hi there.

I just tested out some changes on a preview deployment and the serverless function seems to be crashing and I have no idea why.

<img width="881" alt="Screenshot 2023-06-21 at 16 01 41" src="https://github.com/netlify/next-runtime/assets/8820342/4383e640-f743-4632-aa01-2f2d723c8059">

```
Jun 21, 03:59:26 PM: 2023-06-21T13:59:26.108Z	undefined	ERROR	Uncaught Exception 	{"errorType":"Error","errorMessage":"Could not find Next.js server","stack":["Error: Could not find Next.js server","    at Object.<anonymous> (/var/task/apps/bikeworld/.netlify/functions-internal/___netlify-handler/___netlify-handler.js:2:9)","    at Module._compile (node:internal/modules/cjs/loader:1196:14)","    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)","    at Module.load (node:internal/modules/cjs/loader:1074:32)","    at Function.Module._load (node:internal/modules/cjs/loader:909:12)","    at Module.require (node:internal/modules/cjs/loader:1098:19)","    at require (node:internal/modules/cjs/helpers:108:18)","    at Object.<anonymous> (/var/task/___netlify-handler.js:1:18)","    at Module._compile (node:internal/modules/cjs/loader:1196:14)","    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)"]}
```

https://14-134258--bikeworldmvp.netlify.app/.netlify/functions/___netlify-handler

What more information do you need to help me figure out why this is happening?

**Reply by ben519** [reply]
If my memory serves me correctly, I encountered an error like this too, but my app worked fine locally. So, I tried deploying to Vercel and it worked.

**Reply by j-hit** [reply]
Yep, switching to Vercel also fixed this issue for us.

**Comment by mushfiqueyeasir** [comment]
instead of showing 404 its showing me  this
![screencapture-hr-houze-netlify-app-refund-2023-06-26-05_10_44](https://github.com/netlify/next-runtime/assets/92722825/563b095d-a492-4cf3-9736-908cb953b576)

**Reply by orinokai** [reply]
sorry you are running into this issue, @mushfiqueyeasir, we have a fix rolling out from Monday https://github.com/netlify/next-runtime/pull/2188

**Reply by maxizipitria-halo** [reply]
Was this fixed? I'm getting the same error but the message refers to react instead of react-dom.

**Comment by wrappixelTeam** [comment]
Same issue of page reloading with app routing

**Comment by DWboutin** [comment]
<img width="295" alt="Screenshot 2023-08-22 at 12 45 51 PM" src="https://github.com/netlify/next-runtime/assets/1322006/24ad8456-7f79-4cbb-bd86-a1d74815929d">

Got error 500 when refreshing a `shop/[bundleId]` page on this minimal reproduction repo: https://github.com/DWboutin/test-routing

**Reply by itsjmendez** [reply]
I have the same issue! have you found any solutions?

**Reply by levino** [reply]
I use vercel for the time being. No problems there.

**Reply by itsjmendez** [reply]
Thanks, that's what I ended up doing. I didn't encounter any issues with Vercel.

**Reply by DWboutin** [reply]
We switched on Vercel too, no problems and even faster loads.

**Reply by josip-fundl** [reply]
Any updates or solutions which do not include migrating to Vercel?

**Reply by suhayp-gamma** [reply]
Any solutions!!

**Reply by josip-fundl** [reply]
There is a very very very long discussion held on the Netlify forum. I'll just link the solution which worked for me: https://answers.netlify.com/t/server-edge-not-defined-error-on-nextjs-ssr-functions-cause-site-to-return-500-errors/91793/127

In a nutshell:
1. `npm install next@13.4.9`
2. `npm install @netlify/plugin-nextjs@4.40.1`
3. add the experimental `appDir: true` flag to `next.config,js`

```ts
module.exports = {
  experimental: {
    appDir: true,
  },
}
```

Let me know if this works for you 👏🤞

**Reply by suhayp-gamma** [reply]
Updated Next.js to last version 13.5.2, it is fixed.🚀

**Reply by floka94** [reply]
@suhayp-gamma what is fixed? I am using 13.5.2 and it is not working without installing any plugin.

**Comment by camxus** [comment]
Static Images wont load on Netlify when they load on Vercel properly

<img width="1148" alt="image" src="https://github.com/netlify/next-runtime/assets/27239831/e6bb84ed-08de-4eac-b5ef-343b8a19cb3f">

**Comment by suhayp-gamma** [comment]
Updated Next.js to last version 13.5.2, it is fixed.🚀

**Comment by floka94** [comment]
I am trying to deploy Next.js 13.5.2 App but still get the following:

![image](https://github.com/netlify/next-runtime/assets/29139420/71f2d5c2-738e-4f15-8645-a109dbda5d1c)

I am using the app router and tried to deploy with GitLab integration. In the deploy logs in Netlify everything was "green" and complete.

**Reply by floka94** [reply]
@ascorbic could you fix it?

**Comment by Antonytm** [comment]
Does anybody know if `dynamicParams` still is not supported?

**Reply by FiveOFive** [reply]
Seems like it is not supported yet. Didn't work for me as of today.

I needed `dynamicParams = false` to return 404s for dynamic pages that are only SSG (no ISR). I was able to work around it by having my data fetching functions return null at runtime (as opposed to return the real data at build time for SSG) and the pages handle null and redirect to 404.

```
if (data === null) {
    redirect('/404.html');
  }
```

It's definitely not ideal. The ISR lambda function still gets invoked, but it gives the behavior I needed from dynamicParams
